### ISSUE TITLE
HomeKit two-way audio not working

### ISSUE THREAD

**Issue by longzheng** [koush/scrypted#519]
I was having trouble with getting HomeKit two-way audio working with my ONVIF doorbell. After much debugging and experimenting (thanks to the [amazing plugin development experience](https://github.com/koush/scrypted#debug-scrypted-plugins-in-vscode) that made it so much easier), I've narrowed down two problems with the [Homekit plugin](https://github.com/koush/scrypted/tree/main/plugins/homekit) which I patched locally and it now works perfectly.

I'm sharing my findings here so perhaps these could be fixed upstream in the Homekit plugin with a more official fix and hopefully fix the problem if anyone else is running into them too.

My configuration/environment:
- Server: Windows 10 (with firewall disabled)
- Scrypted server address: 192.168.1.2
- Scrypted version: 0.5.11
- Homekit plugin version: 1.1.91

### Problem 1: SDP return audio FFmpeg port bind error
The first issue I noticed when I tried to active two-way audio is that FFmpeg was throwing an `bind failed: Error number -10048 occurred` error while trying to connect to the SDP return audio server. I saw this in my ONVIF camera console

```
intercom playing
[udp @ 0000020a343f2e00] bind failed: Error number -10048 occurred

[udp @ 0000020a343f2540] bind failed: Error number -10048 occurred
tcp://127.0.0.1:51150: Invalid data found when processing input

ffmpeg exited
```

After extensively reading what other `homebridge` plugins do with two-way audio, I noticed something perculiar in the `homebridge-unifi-protect` code, it [reserves 2 ports for the incoming RTP](https://github.com/hjdhjd/homebridge-unifi-protect/blob/ea1c878b621c5e4bbb6cf1c0f32ce8cd70793b9e/src/protect-stream.ts#L264).

Since `reservePorts` was an utility function in `homebridge`, I then [found a relevant comment on why they do this](https://github.com/homebridge/camera-utils/blob/master/src/ports.ts#LL5-L6)
```ts
// Need to reserve ports in sequence because ffmpeg uses the next port up by default.  If it's taken, ffmpeg will error
export async function reservePorts({
```

The [SDP spec](https://www.ietf.org/rfc/rfc3605.txt) also seems to imply that the consequtive port behaviour is assumed.

> When a session requires multiple ports, SDP assumes that these ports have consecutive numbers.

It's possible other people haven't run into this before by chance, but my server has lots of other services running besides Scrypted so I have a lot of TCP and UDP ports already in-use, so I run into collisions more often.

**Workaround**
In my local environment, I imported the Homebridge `reservePorts` function and changed https://github.com/koush/scrypted/blob/main/plugins/homekit/src/rtp/rtp-ffmpeg-input.ts#L39
```diff
export async function startRtpSink(socketType: SocketType, address: string, srtp: Buffer, audioInfo: AudioInfo, console: Console) {
    const sdpIpVersion = socketType === "udp6" ? "IP6 " : "IP4";
-   const rtpPort = await pickPort();
+   const rtpPort = (await reservePorts({count: 2}))[0];
```

**Potential solutions**
- reserve two non-consequtive ports and explicitly tell FFmpeg to use the two separate ports (I'm not familiar with FFmpeg to know if this is possible)
- port code to reserve/pick two consequtive ports from homebridge

### Problem 2: Audio return buffer not sending to RTP sink

After solving that issue, then I saw FFmpeg was successfully connecting to the SDP server, but then there wasn't anything in the stream.

```

intercom playing
Input #0, sdp, from 'tcp://127.0.0.1:53478':
  Metadata:
    title           : HomeKit Audio Talkback
  Duration: N/A, bitrate: N/A

  Stream #0:0: Audio: opus, 48000 Hz, stereo, s16

Stream mapping:
  Stream #0:0 -> #0:0 (opus (libopus) -> pcm_mulaw (native))
Press [q] to stop, [?] for help

tcp://127.0.0.1:53478: Unknown error

Output #0, rtp, to 'rtp://192.168.1.3:49269?localrtpport=63816&localrtcpport=63817':
 ...
  Stream #0:0: Audio: pcm_mulaw, 8000 Hz, mono, s16, 64 kb/s
    Metadata:
      encoder         : Lavc59.18.100 pcm_mulaw
size=       0kB time=00:00:00.00 bitrate=N/A speed=   0x    
size=       0kB time=00:00:00.00 bitrate=N/A speed=   0x    
video:0kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: unknown
Output file is empty, nothing was encoded (check -ss / -t / -frames parameters if used)

video/audio detected, discarding further input
ffmpeg exited
```

I noticed this was the line that was responsible for sending the audio return buffer into the RTP sink. I added an error callback
```diff
-session.audioReturn.send(buffer, rtpSink.rtpPort);
+session.audioReturn.send(buffer, rtpSink.rtpPort, (error => { error ? console.error("error proxying audio return buffer to RTP sink: " + error.message) : null }));
```

and then saw this in my logs
```
[Intercom with doorbell] error proxying audio return buffer to RTP sink: send EADDRNOTAVAIL undefined:11752
```

It turns out because the `audioReturn` socket was binded with my server address override (`192.168.1.2`)

https://github.com/koush/scrypted/blob/0e0ca55605bfe699d480b856f5a4733153b112ff/plugins/homekit/src/types/camera/camera-streaming.ts#L75

the `.send()` would fail unless it also specified the same address. 

**Workaround**
So I manually edited https://github.com/koush/scrypted/blob/main/plugins/homekit/src/types/camera/camera-streaming.ts#L394 to

```diff
-                       session.audioReturn.send(buffer, rtpSink.rtpPort);
+                       session.audioReturn.send(buffer, rtpSink.rtpPort, '192.168.1.2');
```

and I noticed FFmpeg started picking up the audio stream

```
  Stream #0:0: Audio: pcm_mulaw, 8000 Hz, mono, s16, 64 kb/s
    Metadata:
      encoder         : Lavc59.18.100 pcm_mulaw
size=       0kB time=00:00:00.01 bitrate=  81.8kbits/s speed=N/A    
video/audio detected, discarding further input
```

**Potential solutions**
Add the server address override to the `.send()` function. I ran out of time to figure out a way to refactor the code to get the address override https://github.com/koush/scrypted/blob/main/plugins/homekit/src/types/camera/camera-streaming.ts#L62-L72 into this function.

I would also suggest adding in a console log to the `.send()` callback so that if anyone else runs into issues with the return channel proxy then it will be more easily discoverable.

### Side note: SDP TCP server
Not a problem but I noticed most other plugins don't set up a [separate TCP server for the SDP](https://github.com/koush/scrypted/blob/main/plugins/homekit/src/rtp/rtp-ffmpeg-input.ts#L137-L141) and point FFmpeg to it, instead they all just [pipe the SDP string into FFmpeg `stdin` directly](https://github.com/hjdhjd/homebridge-unifi-protect/blob/ea1c878b621c5e4bbb6cf1c0f32ce8cd70793b9e/src/protect-stream.ts#L769-L770).

I'm not sure if there's any downside to piping but one obvious benefit I see is one less port/socket to open. Not sure if that's something that's been considered before.

**Comment by koush** [comment]
Thanks, appreciate this. This audio return code is absolutely ancient and I've been meaning to rewrite it to take advantage of all the other pipeline gains:

particularly, using the custom RTSP server to restream the data. That will eliminate the sdp, and also allow for multiplexing and better session management/debugging.

**Comment by koush** [comment]
cfc92fd6

**Comment by longzheng** [comment]
> published as beta. uses the rtsp server now. gets rid of all the old code. [cfc92fd](https://github.com/koush/scrypted/commit/cfc92fd65bdd014f06226890ebab551bf343ce48)

Thanks, the beta seems to work perfectly. Appreciate the quick turnaround.

Let me know if there's anything else you'd like me to test regarding this.

I noticed the new code didn't seem to handle `.once('close')` to stop the intercom like the old code did, but assume it's still WIP.

```
                                    session.audioReturn.once('close', () => {
                                        console.log('Stopping intercom.');
                                        device.stopIntercom();
                                    });
```

**Comment by koush** [comment]
> I noticed the new code didn't seem to handle `.once('close')` to stop the intercom like the old code did, but assume it's still WIP.

fixed 0da22066

**Comment by longzheng** [comment]
Presume you mean this commit actually https://github.com/koush/scrypted/commit/cf483bb169b8116401407d4e2a7ca73362298917
